pkg/controller/restore-operator/s3_backup.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2017 The etcd-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"k8s.io/client-go/kubernetes"
+
+	api "github.com/coreos/etcd-operator/pkg/apis/etcd/v1beta2"
+	"github.com/coreos/etcd-operator/pkg/backup"
+	"github.com/coreos/etcd-operator/pkg/backup/backend"
+	"github.com/coreos/etcd-operator/pkg/backup/backupapi"
+	backupS3 "github.com/coreos/etcd-operator/pkg/backup/s3"
+	"github.com/coreos/etcd-operator/pkg/cluster/backupstorage"
+	"github.com/sirupsen/logrus"
+)
+
+const tmpdir = "/tmp"
+
+// backupServer wraps around backup.BackupServer with a close method
+// that cleans up resources associated with backup.BackupServer.
+type backupServer struct {
+	*backup.BackupServer
+	awsDir    string
+	backupDir string
+}
+
+// close removes aws credential/config dir and backupDir dir which were
+// created for BackupServer.
+func (bsc *backupServer) close() {
```

Comment:
probably just remove.

Reply:
the method is called `close` to indicate that this `backupServer` is finished and all its related resources are released.

Reply:
I see..
@xiang90 
So the workflow here is that:
- when a Restore spec is created, it creates S3 client
- When finished serving backup, it destroys everything

The above two is concurrent, although they have implicit ordering.

I am fine with what it is now.

Reply:
Actually after a second thought we shouldn't destroy it. A client might fail before and retry again.
We should keep it there for now. Let's handle cleanup later.

Reply:
After offline discussion, we will save a map of existing Restore CR. Restore CR will be cleaned up once user deletes the CR.

On serveBackup(), we will pull the secret, create storage backend, pull backup, serve backup, and destroy. This eliminates concurrency issue, but might have performance issue. A tradeoff for simplicity now.